node/impl/full/eth.go

```diff
@@ -888,7 +890,12 @@ func (a *EthModule) EthTraceBlock(ctx context.Context, blkNum string) ([]*ethtyp
 			return nil, xerrors.Errorf("when processing message %s: %w", ir.MsgCid, err)
 		}
 
-		err = buildTraces(env, []int{}, &ir.ExecutionTrace)
+		var deployed map[abi.ActorID]*cid.Cid
+		if ir.MsgRct.ExitCode.IsSuccess() {
+			deployed = deployedCode
+		}
```

Comment:
We need to do this per sub-trace. The top-level message can succeed while sub-messages cain fail.

Basically, move this into `buildTraces` and, if the receipt of the currently processed message has a non-zero exit code, set this to nil for the sub-call.